akka-typed/src/main/scala/akka/typed/javadsl/BehaviorBuilder.scala

```diff
@@ -0,0 +1,51 @@
+/**
+ * Copyright (C) 2009-2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.typed.javadsl
+
+import scala.annotation.tailrec
+
+import akka.japi.pf.FI.Apply2
+
+import akka.typed
+import akka.typed.{ Behavior, ExtensibleBehavior, Signal }
+
+case class BehaviorBuilder[T](
+  messageHandlers: List[(Class[_ <: T], (ActorContext[T], T) ⇒ Behavior[T])],
+  signalHandlers:  List[(Class[_ <: Signal], (ActorContext[T], Signal) ⇒ Behavior[T])]
+) {
+
+  def build(): Behavior[T] = {
+    new ExtensibleBehavior[T] {
+      override def receiveMessage(ctx: typed.ActorContext[T], msg: T): Behavior[T] = receive[T](ctx, msg, messageHandlers)
+
+      override def receiveSignal(ctx: typed.ActorContext[T], msg: Signal): Behavior[T] = receive[Signal](ctx, msg, signalHandlers)
+
+      @tailrec
+      private def receive[M](ctx: ActorContext[T], msg: M, handlers: List[(Class[_ <: M], (ActorContext[T], M) ⇒ Behavior[T])]): Behavior[T] =
+        handlers match {
+          case Nil ⇒
+            // emulate scala match error
+            throw new MatchError(msg)
+          case (cls, impl) :: tail ⇒
```

Comment:
Please sanity check if this doesn't allocate an additional tuple in the deconstructing. Not entirely sure if 2.12 handles this nicely (forgot if this is fixed by default nowadays)